lib/internal/streams/operators.js

```diff
@@ -0,0 +1,104 @@
+'use strict';
+
+const { AbortController } = require('internal/abort_controller');
+const { AbortError } = require('internal/errors');
+const Readable = require('internal/streams/readable');
+const eos = require('internal/streams/end-of-stream');
+
+function noop () {}
+
+module.exports.map = function map(stream, fn, options) {
+  // TODO: Argument validation
+
+  let concurrency = 1;
+  if (Number.isFinite(options)) {
+    concurrency = options;
+  } else if (options && Number.isFinite(options.concurrency)) {
+    concurrency = options.concurrency;
+  }
+
+  const ac = new AbortController();
+  const signal = ac.signal;
+  const queue = [];
+
+  let reading = false;
+
+  const ret = new Readable({
+    objectMode: true,
+    highWaterMark: 0,
+    signal: options && options.signal,
+    read () {
+      if (!reading) {
+        read();
+      }
+    },
+    destroy (err, callback) {
+      if (!err && !this.readableEnded) {
+        err = new AbortError();
+      }
+      stream.destroy(err);
+      ac.abort();
+      callback(err);
+    }
+  });
+  const state = ret._readableState;
+
+  async function read () {
+    try {
+      reading = true;
+      while (
+        queue.length &&
+        !state.destroyed &&
+        (!state.length || state.length < state.highWaterMark)
+      ) {
+        let val = queue.shift();
+        if (val && typeof val.then === 'function') {
+          val = await val;
+        }
+        ret.push(val);
+        pump();
+      }
+      reading = false;
+    } catch (err) {
+      ret.destroy(err);
+    }
+  }
+
+  function enqueue(val) {
+    queue.push(val);
+    if (!reading && state.length < state.highWaterMark) {
+      read();
+    }
+  }
+
+  function pump () {
+    try {
+      while (queue.length < concurrency) {
+        let val = stream.read();
+        if (val === null) {
+          return;
+        }
+
+        val = fn(val, { signal });
```

Comment:
I'm wondering if the code can be written in a way that `.filter` "lifts" it for usage since it'd effectively be the same code except these 4 lines in `pump`?

(That can be done at a future date of course but this whole buffering/concurrency operation seems rather generic)

Reply:
Let's do that in next step?

Reply:
Yes, that sounds reasonable.